File: library/two_add/two_add.go
package two_add

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var (
		res  = ListNode{}
		curr = &res
		c    int
	)

	for {

		var v1, v2 int

		if l1 != nil {
			v1 = l1.Val
		}
		if l2 != nil {
			v2 = l2.Val
		}

		val := v1 + v2 + c
		c = 0

		// if need to carry, adjust current and bring carry over to next iteration.
		if r := val % 10; val > 9 {
			c = 1
			val = r
		}

		curr.Val = val

		// we are done when we have no nexts left.
		if (l1 == nil || l1.Next == nil) && (l2 == nil || l2.Next == nil) {
			// in case we are done but we have a carry, we create another node with it as a value
			if c > 0 {
				n := &ListNode{Val: c}
				curr.Next = n
				curr = curr.Next
			}
			break
		}

		curr.Next = &ListNode{}
		curr = curr.Next

		if l1 != nil {
			l1 = l1.Next
		}
		if l2 != nil {
			l2 = l2.Next
		}
	}

	return &res
}

File: library/two_add/two_add.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @Time    : 2020/3/21 下午3:52
# @Author  : serendipity-xp
# @FileName: two_add.py
# @Software: PyCharm
# @GitHub  ：https://github.com/serendipity-xp


# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        dummy_head = ListNode(0)
        cur_l1, cur_l2 = l1, l2
        cur_node = dummy_head
        carry = 0  # 表示进位
        while cur_l1 is not None or cur_l2 is not None:
            x = (cur_l1.val if cur_l1 else 0)
            y = (cur_l2.val if cur_l2 else 0)
            sum_val = x + y + carry
            carry = sum_val / 10
            cur_node.next = ListNode(sum_val % 10)
            cur_node = cur_node.next
            cur_l1 = (cur_l1.next if cur_l1 else None)
            cur_l2 = (cur_l2.next if cur_l2 else None)
        if carry > 0:
            cur_node.next = ListNode(carry)
        return dummy_head.next


if __name__ == '__main__':
    l1 = ListNode(2)
    l1.next = ListNode(4)
    l1.next.next = ListNode(3)
    l2 = ListNode(5)
    l2.next = ListNode(6)
    l2.next.next = ListNode(4)
    Solution().addTwoNumbers(l1, l2)


